doc/design/bucketpath.rst

```diff
@@ -0,0 +1,485 @@
+        protocol = ['bfs', 'bfss']
+
+        def __init__(bucket: Bucket, path: str):
+        """
+        Creates a new BucketPath.
+
+        Args:
+            bucket: used for accssing and interacting with to the underlying bucket.
+            path: of the file within the bucket.
+        """
+            
+        # Pathlike functionalities
+        # ...
+
+
+    # Attention: needs to provide/implment Pathlike protocol
+    class LocalPath:
+        """
+        Provides access to a bucket path served local file system paths.
+        """
+
+        # uri protocol specifies associated with this class
+        protocol = ['bfsl']
+
+        def __init__(path):
+        """
+        Creates a new LocalPath.
+
+        Args:
+            path: of the file within the local file system.
+        """
+
+        # Pathlike functionalities
+        # ...
+
+Modifiers (Extensions)
+----------------------
+
+Modifiers encapsulate exactly one aspect, e.g. changing the root of a Pathlike object.
+
+Each modifier:
+- Modifiers do not have a protocol indicator
+- Extends the URI with on option and value which can be used to reconstruct the Modifier.
+- Must be designed to consider compatibility with other modifiers, potentially working in combination.
+- Must support the ``Pathlike`` interface/protocol
+
+.. code-block:: python
+
+    # Attention: needs to provide/implment Pathlike protocol
+    class Chroot:
+        """
+        Modifies a pathlike object so it will be locked into a specified root.
+        """
+        
+        def __init__(self, path: Pathlike, chroot='/'):
+        """
+        Create a new Chroot.
+
+
+        Args:
+            path: like object which shall be locked into a new root.
+            chroot: the path like object shall be rectricted/pinned to.
+
+        Returns:
+            Creates a new Chroot protected path like object.
+        """
+        # Pathlike functionalities
+        # Note: functions like ``parent`` should stop at the new root
+        # ...
+
+
+    # Attention: needs to provide/implment Pathlike protocol
+    class ReadOnly:
+        """
+        Modifies a pathlike object so it will be readonly.
+        """
+
+        def __init__(path: Pathlike):
+            """
+            Create a new ReadOnly Pathlike object.
+
+            Args:
+                path: like object which will be write proected (readonly).
+
+
+            Returns:
+                A path like object whith write proection.
+            """
+        
+        # Pathlike functionalities
+        # Note: Non readonly actions should throw an exception
+        # ...
+
+
+
+Factory & Builders
+------------------
+
+.. code-block:: python
+
+    def PathBuilder:
+
+        def __init__(credentails_store, *args, **kwargs):
+            """
+            Initalizes the factory with settings which are required besides the actual path uri.
+
+            Args:
+
+                credentail_store: for accessing buckets, see Service.
```

Comment:
the credential_store probably will get later, a bit criminal, because you might remember, I that,that we have a use case, where I need to wipe all the credential information from the Bucket Objects/Pathlike Objects before serializing them and later reinject the credentials in the deserialized Bucket Objects/Pathlike without that the part that serializes and deserialized has direct references to the Bucket Objects/Pathlike 


For example:
```
def step1(credentials)
    credential_store = CredentialStore(credentials)
    Path = PathFactory(credential_store, ...)
    component = Component(Path)
    component.do_something() -> this creates PathLikes internallly which get save in the component
    credential_store.clean()
    serialize(component, credential_store)

def step2(credentials)
    component = deserialize()
    component, credential_store = CredentialStore(...)
    credential_store.repopulate(credentials)
    component.do_something_else() -> uses the stored PathLikes
  


```

Reply:
From my pov the serialization of the `Paths` do not carry any credentials:
```python
https_bucket_path = Path("bfss://127.0.0.1:8888/service/bucket/some/file.tar.gz")
```

Still even if we run in such a scenario, as discussed back than I think the `Credential Store` abstraction can take care of this.